zephyr: Upgrade to Zephyr v3.7.0.

MaureenHelm opened the pull request:
Updates the Zephyr port build instructions and CI to use the latest
Zephyr release tag.

Tested on frdm_k64f.

~~In draft status until the Zephyr release is finalized (currently in release candidate / stabilization phase).~~
[Zephyr v3.2.0](https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.2.0) was released on Sep 30 2022.
[Zephyr v3.3.0](https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.3.0) was released on Feb 19 2023.
[Zephyr v3.4.0](https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.4.0) was released on Jun 16 2023.
[Zephyr v3.5.0](https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.5.0) was released on Oct 20 2023.
[Zephyr v3.6.0](https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.6.0) was released on Feb 23 2024.
[Zephyr v3.7.0](https://github.com/zephyrproject-rtos/zephyr/releases/tag/v3.7.0) was released on Jul 26 2024.

Fixes #9115

MaureenHelm committed:
zephyr: Replace zephyr.h with kernel.h.

Zephyr v3.2.0 deprecated include/zephyr/zephyr.h in favor of
include/zephyr/kernel.h since it only included that header.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
docs/zephyr: Update docs to reflect device name changes.

Zephyr v3.2.0 deprecated the devicetree label property as a base
property, which had been used as the device name string for
device_get_binding(). The device name string is now the devicetree node
name appended with its unit-address. Update Zephyr port documentation
to reflect this change.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
tools/ci.sh: Upgrade Zephyr docker image 0.26.13 and SDK 0.16.8.

Upgrades CI to use the latest versions of the Zephyr docker image and
Zephyr SDK.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
zephyr: Upgrade to Zephyr v3.7.0.

Updates the Zephyr port build instructions and CI to use the latest
Zephyr release tag.

Tested on frdm_k64f.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
zephyr: Update include paths to use the zephyr namespace.

Zephyr v3.1.0 moved all public headers to include/zephyr. Updates a few
Zephyr include paths that were missed in
4fd54a475661953946e9e3c3f962b20619c82240.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

MaureenHelm committed:
zephyr: Replace deprecated FLASH_AREA macros with FIXED_PARTITION.

Zephyr v3.2.0 deprecated FLASH_AREA macros in favor of FIXED_PARTITION
macros, using node labels instead of node label properties to reference
flash storage partitions.

Signed-off-by: Maureen Helm <maureen.helm@intel.com>

dpgeorge commented on docs/zephyr/quickref.rst:
```diff
@@ -36,7 +36,10 @@ Use the :ref:`machine.Pin <machine.Pin>` class::
 
-    pin = Pin(("GPIO_1", 21), Pin.IN)   # create input pin on GPIO1
+    gpio1 = "gpio@400ff040"             # GPIO1 device name
+    gpio2 = "gpio@400ff080"             # GPIO2 device name
```
This seems quite.... un-user-friendly, to know the address.  Is there a better solution?  How do other Zephyr projects handle this?  Maybe in the device tree / overlay we can have an alias?

dpgeorge replied:
As @jimmo says below, he'll look into this.

MaureenHelm replied:
Please have a look at #10859

jimmo commented:
Thanks @MaureenHelm !!

I think we're going to need to come up with a solution for the issue discussed in https://github.com/micropython/micropython/issues/9115 before merging this, i.e. we need to make it possible to instantiate pins and other peripherals without needing to know the register base addresses. I will investigate.

bogdanm commented:
Adding a few thoughts to this discussion: we might be able to use some of the build files generated by Zephyr to solve this problem. For example, there is a `build/zephyr/dts.cmake` file that looks like this:

```
...
set_target_properties(devicetree_target PROPERTIES "DT_NODELABEL|gpio0" "/soc/peripheral@50000000/gpio@842500")
set_target_properties(devicetree_target PROPERTIES "DT_NODELABEL|gpio1" "/soc/peripheral@50000000/gpio@842800")
...

```

This can be parsed by a Python script to generate more friendly names (`gpio0` or `GPIO_0` instead of `gpio@842500`) for the various GPIO ports and possible other DT nodes (a cmake wizard might be able to do this directly from cmake, but a cmake wizard I am not). I'll check which Zephyr script(s) are used to generate this file, we might be able to use them.

There's also `build/zephyr/zephyr.dts` which can be used in the same way, not sure which one is easier (cmake parsing for the above file would be pretty easy to do in Python, but I'm guessing there are some Python packages that can parse .dts, so the latter option might be easier after all). 

The point is that we probably need this additional processing step and it needs to be plugged in into Zephyr's cmake machinery. I'm not sure how to do this yet, but I'm going to look deeper into it, because I don't see a better way to fix this.

MaureenHelm committed:
zephyr: Change main function to return an int.

Zephyr v3.4.0 changed the declaration of the main function to return an
int to allow building Zephyr without the -ffreestanding compiler flag.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm committed:
zephyr: Change SPI chip select from a pointer to a struct member.

Zephyr v3.4.0 changed the SPI chip select from a pointer to a struct
member to allow using the existing SPI dt-spec macros in C++.

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm commented:
Rebased and updated for Zephyr v3.4.0. The issue discussed in #10859 and #9115 still remains.

jimmo commented:
> Rebased and updated for Zephyr v3.4.0. The issue discussed in #10859 and #9115 still remains.

I have a resolution to both these working, based on the idea demonstrated by @bogdanm , just haven't found the time this week to get the PR finalised.

MaureenHelm commented:
Rebased and updated for Zephyr v3.5.0

MaureenHelm commented:
Rebased and updated for Zephyr v3.6.0

danicampora commented:
> > Rebased and updated for Zephyr v3.4.0. The issue discussed in #10859 and #9115 still remains.
> 
> I have a resolution to both these working, based on the idea demonstrated by @bogdanm , just haven't found the time this week to get the PR finalised.

@jimmo I would love to help you with this! I'm very interested in getting MicroPython to work with the latest Zephyr and also my the goal is to extend the features of the Zephyr port, specially for Nordic devices. Please let me know how I could join forces with you to get this over the line :-)

Rafeeq-Muhammad commented:
I support this :)

ZbynekFedra commented:
Like to see an activity on this pull request. I ma also voting for this merge. We have some projects to be ideally build on top of MicroPython on Zephyr port.

MaureenHelm committed:
zephyr: Allow using devicetree node labels to construct machine objects.

Zephyr v3.7.0 added a new feature to allow getting devices by their
devicetree node labels. Use this feature in the MicroPython Zephyr port
to simplify constructing machine module objects, including Pin, SPI,
I2C, and UART. It's still possible to use the more verbose device names
(e.g., gpio@400ff040, i2c@40066000, spi@4002c000), but now we can also
use their devicetree node labels (e.g., gpiob, i2c0, spi0).

Node labels aren't standardized across all SoC families because they
generally try to follow their respective SoC hardware user manual naming
convention, however many boards define common labels for devices routed
to Arduino headers (e.g., arduino_i2c, arduino_serial, and arduino_spi).
That means I2C("arduino_i2c") will work on quite a few boards (>100 in
the main Zephyr tree).

Signed-off-by: Maureen Helm <maureen.helm@analog.com>

MaureenHelm commented:
Rebased and updated for Zephyr v3.7.0.

A new feature was introduced in zephyrproject-rtos/zephyr#71912 that helps to address the issue discussed in #10859 and #9115. Please take a look at 207a4a320f17e235e680937e571c0cf6d0dbfaec.

MaureenHelm commented:
> Rebased and updated for Zephyr v3.7.0.
> 
> A new feature was introduced in [zephyrproject-rtos/zephyr#71912](https://github.com/zephyrproject-rtos/zephyr/pull/71912) that helps to address the issue discussed in #10859 and #9115. Please take a look at [207a4a3](https://github.com/micropython/micropython/commit/207a4a320f17e235e680937e571c0cf6d0dbfaec).

@dpgeorge ping

danicampora commented:
@MaureenHelm wow, nice job!
@dpgeorge I'm very keen on having MicroPython working with the latest Zephyr. If there's anything I can do to help with this pull request please let me know :-)

MaureenHelm commented:
Rebased and fixed a merge conflict

dpgeorge commented:
Thank you very much @MaureenHelm for maintaining this PR for nearly two years.

The latest changes to support specifying a device by its DT node label look really good, that's pretty much exactly what we were looking for.  I think that addresses all three of @jimmo 's points in https://github.com/micropython/micropython/pull/10859#issuecomment-1447653246:
1. No need to maintain any special board configurations/overlays to get the user-friendly device names.
2. Works very similar to other ports, the user just specifies a string describing the peripheral name, and no need for a device address.
3. Someone coming from Zephyr to MicroPython, they can easily transfer across their knowledge about DT names, if they know the DT node label then they know how to access a peripheral.

Independent to the above points, at the very least this PR is now much less of a user-facing change: the changes that the user sees are that device/peripheral names have slightly changed, eg `"I2C_0"` is now `"i2c0"`.  I think that's a pretty good outcome overall.

dpgeorge commented on docs/zephyr/quickref.rst:
```diff
@@ -36,7 +36,10 @@ Use the :ref:`machine.Pin <machine.Pin>` class::
 
-    pin = Pin(("GPIO_1", 21), Pin.IN)   # create input pin on GPIO1
+    gpiob = "gpiob"                     # GPIO port B devicetree node label
+    gpioc = "gpioc"                     # GPIO port C devicetree node label
```
Is it worth defining variables like this?  The original example code had the strings written inline in the `Pin` constructor, and I think that's slightly better from an example point of view because it makes it easy to just cut-and-paste a single line and it should just work.  Eg:
```py
pin = Pin(("gpiob", 21), Pin.IN)
```
That's obvious at a glance just at that line how to specify a pin name.

dpgeorge commented on ports/zephyr/machine_i2c.c:
```diff
@@ -67,6 +67,12 @@ mp_obj_t machine_hard_i2c_make_new(const mp_obj_type_t *type, size_t n_args, siz
+        dev = device_get_by_dt_nodelabel(dev_name);
+    }
+    #endif
+
```
I suggest making a general helper function for this bit:
```c
const struct device *zephyr_device_find(mp_obj_t name) {
    const char *dev_name = mp_obj_str_get_str(name);
    const struct device *dev = device_get_binding(dev_name);

    #ifdef CONFIG_DEVICE_DT_METADATA
    if (dev == NULL) {
        dev = device_get_by_dt_nodelabel(dev_name);
    }
    #endif

    if (dev == NULL) {
        #if MICROPY_ERROR_REPORTING <= MICROPY_ERROR_REPORTING_TERSE
        mp_raise_ValueError(MP_ERROR_TEXT("device not found"));
        #else
        mp_raise_msg_varg(&mp_type_ValueError, MP_ERROR_TEXT("device %s not found"), dev_name);
        #endif
    }

    return dev;
}
```
That can be used in quite a few places in this PR.

MaureenHelm committed:
zephyr: Refactor device lookup into a common helper function.

Refactors Zephyr device lookup operations into a common helper function
to reduce boilerplate code that was repeated in multiple modules.

Suggested-by: Damien George <damien@micropython.org>
Signed-off-by: Maureen Helm <maureen.helm@analog.com>

danicampora commented:
Amazing! Looking forward for this to be merged :-)

MaureenHelm commented:
Anything else needed to get this merged?

MaureenHelm commented:
> @MaureenHelm If you don't mind, during merge I would squash some of the commits here, eg it probably needs only one commit to update the Docker image and SDK in `tools/ci.sh`, and then one commit to update to Zephyr v3.7.0.

Ok

dpgeorge commented:
I've rebased this on latest master, and squashed some of the commits, and force-pushed to this branch.

I was ready to merge it but the CI is not happy:
```
/micropython/py/emitbc.c:926:1: fatal error: error writing to /tmp/ccCLRy5M.s: No space left on device
  926 | };
      | ^
compilation terminated.
```
Looks like the CI VM itself ran out of disk space??  That's only when building boards latter in the test, so maybe we can clean up previous builds to free up space?

There's also the following warning which I think needs to be fixed:
```
/micropython/ports/zephyr/zephyr_device.c: In function 'zephyr_device_find':
/micropython/ports/zephyr/zephyr_device.c:32:32: warning: implicit declaration of function 'device_get_binding' [-Wimplicit-function-declaration]
   32 |     const struct device *dev = device_get_binding(dev_name);
      |                                ^~~~~~~~~~~~~~~~~~
/micropython/ports/zephyr/zephyr_device.c:32:32: warning: initialization of 'const struct device *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
```

dpgeorge committed:
github/workflows: Free up disk space on zephyr workflow.

The zephyr builds take up quite a lot of space.

Signed-off-by: Damien George <damien@micropython.org>

dpgeorge commented:
I fixed the CI out-of-space issue by using https://github.com/marketplace/actions/free-disk-space-ubuntu in the zephyr workflow.

dpgeorge commented:
Merged!

Thanks again @MaureenHelm for keeping this PR up-to-date.

MaureenHelm commented:
> There's also the following warning which I think needs to be fixed:
> 
> ```
> /micropython/ports/zephyr/zephyr_device.c: In function 'zephyr_device_find':
> /micropython/ports/zephyr/zephyr_device.c:32:32: warning: implicit declaration of function 'device_get_binding' [-Wimplicit-function-declaration]
>    32 |     const struct device *dev = device_get_binding(dev_name);
>       |                                ^~~~~~~~~~~~~~~~~~
> /micropython/ports/zephyr/zephyr_device.c:32:32: warning: initialization of 'const struct device *' from 'int' makes pointer from integer without a cast [-Wint-conversion]
> ```

Fixed

> I fixed the CI out-of-space issue by using [github.com/marketplace/actions/free-disk-space-ubuntu](https://github.com/marketplace/actions/free-disk-space-ubuntu) in the zephyr workflow.

Thanks!
